sklearn/cluster/tests/test_hierarchical.py

```diff
@@ -527,13 +528,141 @@ def test_agg_n_clusters():
 
     rng = np.random.RandomState(0)
     X = rng.rand(20, 10)
-    for n_clus in [-1, 0]:
-        agc = AgglomerativeClustering(n_clusters=n_clus)
-        msg = ("n_clusters should be an integer greater than 0."
-               " %s was provided." % str(agc.n_clusters))
-        assert_raise_message(ValueError, msg, agc.fit, X)
+    n_clus = -1
+    agc = AgglomerativeClustering(n_clusters=n_clus)
+    msg = ("n_clusters should be an integer greater than 0."
+           " %s was provided." % str(agc.n_clusters))
+    assert_raise_message(ValueError, msg, agc.fit, X)
+    n_clus = 0
+    agc = AgglomerativeClustering(n_clusters=n_clus)
+    msg = ("Either n_clusters (>0) or distance_threshold "
+           "needs to be set, got n_clusters={} and "
+           "distance_threshold=None instead.".format(agc.n_clusters))
+    assert_raise_message(ValueError, msg, agc.fit, X)
+
+
+def test_agg_n_cluster_or_distance_threshold():
+    # Test that an error is raised when n_clusters
+    # and distance_threshold are None
+
+    n_clus = None
+    rng = np.random.RandomState(0)
+    X = rng.rand(20, 10)
+    # distance_threshold is None by default
+    agc = AgglomerativeClustering(n_clusters=n_clus)
+    msg = ("Either n_clusters (>0) or distance_threshold "
+           "needs to be set, got n_clusters={} and "
+           "distance_threshold=None instead.".format(agc.n_clusters))
+    assert_raise_message(ValueError, msg, agc.fit, X)
+
+
+def test_agg_n_cluster_and_distance_threshold():
+    # Test that when distance_threshold is set
+    # n_clusters_ is set to None
+
+    n_clus, dist_thresh = 10, 10
+    rng = np.random.RandomState(0)
+    X = rng.rand(20, 10)
+    agc = AgglomerativeClustering(n_clusters=n_clus,
```

Comment:
I'm a bit undecided about whether we should silently ignore n_clusters rather than force the user to explicit set it to None. But if we take the approach that setting a distance_threshold means n_clusters is ignored, then we need to state that in the documentation of both parameters.